File: programmers-coding-test/Python/Level3/단어 변환.py
answer = 51


def similar(s1, s2):
    difference = 0
    for i in range(len(s1)):
        if s1[i] != s2[i]:
            difference += 1
    return difference == 1


def dfs(lev, begin, target, words, used):
    global answer
    if begin == target:
        answer = min(answer, lev)
        return
    for i in range(len(words)):
        word = words[i]
        if similar(begin, word) and used[i] == 0:
            used[i] = 1
            dfs(lev + 1, word, target, words, used)
            used[i] = 0


def solution(begin, target, words):
    used = [0] * len(words)
    dfs(0, begin, target, words, used)
    return answer if answer != 51 else 0

File: programmers/C++/Level3/단어 변환.cpp
#include <string>
#include <vector>

using namespace std;

int cnt = 21e8;
int used[20];

bool isSimil(string s1, string s2)
{
	int count = 0;
	for (int i = 0; i < s1.length(); i++)
	{
		if (s1[i] != s2[i]) count++;
	}
	return (count == 1);
}

void dfs(int lev, string now, string target, vector<string> words)
{
	if (now == target && lev < cnt)
	{
		cnt = lev;
		return;
	}
	for (int i = 0; i < words.size(); i++)
	{
		if (used[i] == 0 && isSimil(words[i], now))
		{
			used[i] = 1;
			dfs(lev + 1, words[i], target, words);
			used[i] = 0;
		}
	}
}

int solution(string begin, string target, vector<string> words) {
	int answer = 0;

	dfs(0, begin, target, words);
    if (cnt == 21e8) answer = 0;
	else answer = cnt;

	return answer;
}